go/ai/tools.go

```diff
@@ -30,29 +34,58 @@ const provider = "local"
 // TODO: This should be generic over the function input and output types,
 // and something in the general code should handle the JSON conversion.
 type Tool struct {
-	Definition *ToolDefinition
-	Fn         func(context.Context, map[string]any) (map[string]any, error)
+	Action action.Action
 }
 
 // DefineTool defines a tool function.
-func DefineTool(definition *ToolDefinition, metadata map[string]any, fn func(ctx context.Context, input map[string]any) (map[string]any, error)) {
+func DefineTool[In, Out any](name, description string, metadata map[string]any, fn func(ctx context.Context, input In) (Out, error)) *Tool {
```

Comment:
Why not carry the type safety through as far as possible, by having a `Tool[In, Out]` that contains a `core.Action[In, Out, struct{}]`.

Reply:
At some point you might need a `Tool` type that wasn't generic, as we do for actions, and it would have a RunJSON method. Or maybe you could just use actions for that, I'm not sure.

Reply:
I spent a lot of time trying to make Tool generic... in most Actions (except custom actions) generic i/o types are fixed (ex. model action has GenerateRequest/Response types), so in, for example. LookupModel you can fill in the Action generics (in `LookupActionFor`) with hardcoded types. With Tools the i/o types are defined by the user per tool. So I need a non-generic type to be able to construct and return it from `LookupTool`.

Reply:
actually nevermind... figured it out.